_drafts/learn-doobie-for-great-good.md

```diff
@@ -0,0 +1,961 @@
+
+The last aspect we left about selecting information from a table is to parameterize the query with parameters. Fortunately, the `sql` interpolator works smoothly with parameters, using the exact mechanism used by Scala native `String` interpolation:
+
+```scala
+def findActorsByNameInitialLetterProgram(initialLetter: String): IO[List[Actor]] = {
+  val findActors: fs2.Stream[doobie.ConnectionIO, Actor] =
+    sql"select id, name from actors where LEFT(name, 1) = $initialLetter".query[Actor].stream
+  findActors.compile.toList.transact(xa)
+}
+```
+
+The above program extracts all `actors` from the table whose names start with the given initial letter. As we can see, passing a parameter to a query is as simple as passing it to an interpolated string.
+
+### 3.1. The `HC` Module
+
+The interpolator `sql` is a very handful syntax for writing SQL queries. However, it's not the only way to write queries. In fact, it's just syntactic sugar to access functions available in the `doobie.hi.connection` module, aliased as `HC`.
+
+So, let's take the first program we developed above, the `findAllActorsNamesProgram`, and desugar it using the `HC` module:
+
+```scala
+def findActorByNameUsingHCProgram(actorName: String): IO[Option[Actor]] = {
+  val query = "select id, name from actors where name = ?"
+  HC.stream[Actor](
+    query,
+    HPS.set(actorName),   // Parameters start from index 1 by default
+    512
+  ).compile
+    .toList
+    .map(_.headOption)
+    .transact(xa)
+}
+```
+
+First, the query becomes a plain `String` containing `?` wildcards. The `sql` interpolator is just syntactic sugar for the `HC.stream[A]` method. Leaving the comprehension of the first parameter to the reader, the second parameter has type `PreparedStatementIO[B]`. As for the `ConnectionIO[A]` type, a `PreparedStatementIO` is an instance of the free monad pattern. In this case, it describes how to inject parameters into the query. So, the interpreter of the monad, `HC`, builds a `Kleisli[IO, PreparedStatement, A]`, which is precisely a function returning an `IO[A]` given a `PreparedStatement`.
+
+The third parameter is the maximum number of rows to be fetched at a time. In fact, Doobie read rows in chunks.
+
+If the query has more than one parameter, we have many choices on the syntax to use:
+
+```scala
+// Set parameters as (Int, String)
+HPS.set((1, "Henry Cavill"))
+
+// Set parameters individually
+HPS.set(1, 1) *> HPS.set(2, "Henry Cavill")
+
+// ...and many others!
+```
+
+### 3.1. Fragments
+
+Until now, we used the `sql` interpolator to build our queries. Indeed, the `sql` interpolator is just an alias for the `fr` interpolator, which name stands for `Fragment`. **A fragment is a piece of an SQL statement that we can combine with any other fragment to build a proper SQL instruction**.
```

Comment:
Slight rephrasing:
It turns out that the sql interpolator is an alias of the more general `fr` interpolator, whose name stands for Fragment.